consensus.go

```
@@ -0,0 +1,245 @@
+	wal := wallet.Internal()
+
+	if vspHost == "" && hash == "" {
+		return fmt.Errorf("request requires either a vsp host or a ticket hash")
+	}
+
+	// Test to ensure the passphrase is correct, before setting a voting
+	// preference.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	ctx := wallet.shutdownContext()
+
+	// get choices
+	choices, _, err := wallet.Internal().AgendaChoices(ctx, ticketHash) // returns saved prefs for current agendas
+	if err != nil {
+		return err
+	}
+
+	currentChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: "abstain", // default to abstain as current choice if not found in wallet
+	}
+
+	for i := range choices {
+		if choices[i].AgendaID == agendaID {
+			currentChoice.ChoiceID = choices[i].ChoiceID
+			break
+		}
+	}
+
+	newChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, newChoice)
+	if err != nil {
+		return err
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket
+	// with the associated vsp.
+	if ticketHash != nil {
+		vspTicketInfo, err := wal.VSPTicketInfo(ctx, ticketHash)
+		if err != nil {
+			return err
+		}
+		// Update the vote choice for the ticket with the vsp
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil {
+			return err
+		}
+		err = vspClient.SetVoteChoice(ctx, ticketHash, newChoice)
+		if err != nil {
+			// Updating the agenda voting preference with the vsp failed, therefore
+			// revert the locally saved voting preference for the agenda.
+			_, err = wal.SetAgendaChoices(ctx, ticketHash, currentChoice)
+			if err != nil {
+				return err
+			}
+		}
+
```

Ignore errors from trying to reset the prefs, return the original error.
```suggestion
		err = vspClient.SetVoteChoice(ctx, ticketHash, newChoice)
		if err != nil {
			// Updating the agenda voting preference with the vsp failed,
			// revert the locally saved voting preference for the agenda.
			_, revertError := wal.SetAgendaChoices(ctx, ticketHash, currentChoice)
			if revertError != nil {
				// log the revertError
			}
			return err
		}
```